README.md
# Neonpen

A Flutter package to add a Neon Pen style on text.<br>
There are options for color, opacity, ziggle, double layer and so on.
<br><br>


# Preview
<img src="https://user-images.githubusercontent.com/24351423/194112767-75e38b5b-87c5-4834-b177-4ffaa75dab76.png" width="300">



# Installation

Add `neonpen: ^2.0.5` to your `pubspec.yaml` dependecies. And import it:

```
import 'package:neonpen/neonpen.dart';
```
<br>

# How to use
Simply add a Animation List widget with required params.

```
Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.red,
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 0,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.yellow,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.3,
              disableEmphasis: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.orange,
              padding: EdgeInsets.only(
                top: -10,
                left: 5,
                right: 5,
              ),
              opacity: 0.3,
              disableEmphasis: true,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.grey,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.7,
              emphasisWidth: 5,
              emphasisOpacity: 0.5,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.blue,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.3,
              emphasisWidth: 15,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 2,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.purple,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.3,
              emphasisWidth: 10,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 2,
              enableLineZiggle: true,
              lineZiggleLevel: 2,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.green,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 20),
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.lime,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 20),
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 2,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.amber,
              padding: EdgeInsets.symmetric(vertical: 10, horizontal: 25),
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.indigo,
              padding: EdgeInsets.symmetric(vertical: 10, horizontal: 25),
              opacity: 0.3,
              emphasisWidth: 8,
              emphasisOpacity: 0.1,
              emphasisAngleDegree: 0.8,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
          ],
        ),
      ),
    );
}
```
<br>

# Properties
| Attribute | Data type | Description | Default |
|--|--|--|--|
| text | Text | A Text widget that will be covered by neonpen effect | - |
| color | Color | A color of neonpen style | - |
| opacity | double | An opacity of neonpen style | 0.5 |
| padding | EdgeInsets | A padding of between Text widget and neonpen style | EdgeInsets.symmetric(horizontal: 5) |
| disableEmphasis | bool | A bool value to disable emphasis effect | false |
| emphasisWidth | double | The start and end's neonpen style thickness | 5.0 |
| emphasisOpacity | double | An opacity of the start and end's neonpen style | 0.55 |
| emphasisAngleDegree | double | An angle degree of the start and end's neonpen style | 1 |
| enableLineZiggle | bool | Add a ziggle style to neonpen line | false |
| lineZiggleLevel | double | A level of ziggle style to neonpen line | 1 |
| isDoubleLayer | bool | Add a two neopen style on Text widget | false |


CHANGELOG.md
## [2.0.5] 
* Added disableEmphasis property to disable emphasis effect

## [2.0.4] 
* Changed from StatefulWidget to StatelessWidget
* Changed emphasis line style
* Refactoring structure

## [2.0.2] 
* fix example package path error

## [2.0.1] 
* Add comments
* Changed Example location

## [2.0.0] 
* Initial release


test/neonpen_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:neonpen/neonpen.dart';

void main() {}


lib/neonpen.dart
library neonpen;

import 'dart:math';

import 'package:flutter/material.dart';

// A Widget that give a neon pen style to Text widget
class Neonpen extends StatelessWidget {
  final Text text;
  final Color color;
  final double? opacity;
  final EdgeInsets? padding;
  final bool? disableEmphasis;
  final double? emphasisWidth;
  final double? emphasisOpacity;
  final double? emphasisAngleDegree;
  final bool? enableLineZiggle;
  final double? lineZiggleLevel;
  final bool? isDoubleLayer;

  Neonpen({
    required this.text,
    required this.color,
    this.opacity,
    this.padding,
    this.disableEmphasis,
    this.emphasisWidth,
    this.emphasisOpacity,
    this.emphasisAngleDegree,
    this.enableLineZiggle,
    this.lineZiggleLevel,
    this.isDoubleLayer,
  });

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      foregroundPainter: NeonPainter(
        color: color,
        padding: padding,
        opacity: opacity,
        disableEmphasis: disableEmphasis,
        emphasisWidth: emphasisWidth,
        emphasisOpacity: emphasisOpacity,
        emphasisAngleDegree: emphasisAngleDegree,
        enableLineZiggle: enableLineZiggle,
        lineZiggleLevel: lineZiggleLevel,
        isDoubleLayer: isDoubleLayer,
      ),
      child: text,
    );
  }
}

// A NeonPainter class for draw neon pen style
class NeonPainter extends CustomPainter {
  final double opacity;
  final Color color;
  final EdgeInsets padding;
  final bool disableEmphasis;
  final double emphasisWidth;
  final double emphasisOpacity;
  final double emphasisAngleDegree;
  final bool enableLineZiggle;
  final double lineZiggleLevel;
  final bool isDoubleLayer;

  final _random = Random();

  NeonPainter({
    required Color color,
    double? opacity,
    EdgeInsets? padding,
    bool? disableEmphasis,
    double? emphasisWidth,
    double? emphasisOpacity,
    double? emphasisAngleDegree,
    bool? enableLineZiggle,
    double? lineZiggleLevel,
    bool? isDoubleLayer,
  })  : this.opacity = opacity ?? 0.5,
        this.color = color,
        this.padding = padding ?? EdgeInsets.symmetric(horizontal: 5),
        this.disableEmphasis = disableEmphasis ?? false,
        this.emphasisWidth = emphasisWidth ?? 5.0,
        this.emphasisOpacity = emphasisOpacity ?? (opacity ?? 0.5 + 0.05),
        this.emphasisAngleDegree = emphasisAngleDegree ?? 1,
        this.enableLineZiggle = enableLineZiggle ?? false,
        this.lineZiggleLevel = lineZiggleLevel ?? 1,
        this.isDoubleLayer = isDoubleLayer ?? false;

  @override
  void paint(Canvas canvas, Size size) {
    final double startX = 0 - padding.left;
    final double endX = size.width + padding.right;
    final double startY = 0 - padding.top;
    final double endY = size.height + padding.bottom;

    if (isDoubleLayer == false) {
      // If widget has a one layer, just draw full size neonpen style
      drawNeonpen(
        canvas: canvas,
        x: startX,
        y: startY,
        width: endX,
        height: endY,
      );
    } else {
      // If widget has a double layer, draw one neopen style and
      // over-draw over first neonpen style layer
      drawNeonpen(
        canvas: canvas,
        x: startX,
        y: startY,
        width: endX,
        height: size.height * 2 / 3 + padding.bottom * 3 / 4,
      );
      drawNeonpen(
        canvas: canvas,
        x: startX + 5,
        y: size.height * 2 / 3 - padding.bottom * 3 / 4,
        width: endX + 5,
        height: endY,
      );
    }
  }

  // Draw neonpen at position of x,y and size of height, width.
  void drawNeonpen({
    required Canvas canvas,
    required double height,
    required double width,
    double x = 0,
    double y = 0,
  }) {
    final double startX = x;
    final double endX = width;
    final double startY = y;
    final double endY = height;
    final double leftRightDistance = 3.0 * emphasisAngleDegree;

    // Draw main neon main line
    Paint paint = Paint()
      ..color = color.withOpacity(opacity)
      ..style = PaintingStyle.fill;

    Path path = Path();
    path.addPolygon([
      Offset(startX, startY),
      if (enableLineZiggle)
        ...makeRandomOffset(startX, endX, startY, 50 * lineZiggleLevel.toInt(), 0.3 * lineZiggleLevel),
      Offset(endX, startY),
      Offset(endX - leftRightDistance, endY),
      if (enableLineZiggle)
        ...makeRandomOffset(startX - leftRightDistance, endX - leftRightDistance, endY, 50 * lineZiggleLevel.toInt(),
            0.3 * lineZiggleLevel,
            reverse: true),
      Offset(startX - leftRightDistance, endY),
    ], true);
    canvas.drawPath(path, paint);

    // Draw left, right emphasis line
    if (!disableEmphasis) {
      Paint emphasisPaint = Paint()
        ..color = color.withOpacity(emphasisOpacity)
        ..style = PaintingStyle.stroke
        ..strokeWidth = emphasisWidth
        ..strokeCap = StrokeCap.round;

      Path leftEmphasisPath = Path();
      leftEmphasisPath.moveTo(startX + (emphasisWidth / 2), startY + (emphasisWidth / 2));
      leftEmphasisPath.lineTo(startX + (emphasisWidth / 2) - leftRightDistance, endY - (emphasisWidth / 2));
      canvas.drawPath(leftEmphasisPath, emphasisPaint);

      Path rightEmphasisPath = Path();
      rightEmphasisPath.moveTo(endX - (emphasisWidth / 2), startY + (emphasisWidth / 2));
      rightEmphasisPath.lineTo(endX - (emphasisWidth / 2) - leftRightDistance, endY - (emphasisWidth / 2));

      canvas.drawPath(rightEmphasisPath, emphasisPaint);
    }
  }

  @override
  bool shouldRepaint(covariant NeonPainter oldDelegate) {
    // return oldDelegate.width != this.width ||
    //     oldDelegate.height != this.height ||
    //     oldDelegate.color != this.color ||
    //     oldDelegate.opacity != this.opacity ||
    //     oldDelegate.padding != this.padding ||
    //     oldDelegate.emphasisWidth != this.emphasisWidth;
    return true;
  }

  // To get range of random double value
  double doubleInRange(Random source, num start, num end) => source.nextDouble() * (end - start) + start;

  // Make a list of random Offset
  // startMainAxis : random start value
  // endMainAxis : radom end value
  // amount : amount of random Offset value
  // differenceCrossAxis : gap from crossAxis 0 value
  List<Offset> makeRandomOffset(
      double startMainAxis, double endMainAxis, double startCrossAxis, int amount, double differenceCrossAxis,
      {bool reverse = false}) {
    List<Offset> differeceOffsets = [];

    for (int loop = 0; loop < amount; loop++) {
      double _x = _random.nextDouble() * (endMainAxis - startMainAxis);
      double _y = startCrossAxis + (_random.nextBool() ? -1 : 1) * doubleInRange(_random, 0, differenceCrossAxis);
      differeceOffsets.add(Offset(startMainAxis + _x, _y));
    }

    differeceOffsets.sort((a, b) => reverse ? b.dx.compareTo(a.dx) : a.dx.compareTo(b.dx));

    return differeceOffsets;
  }
}


lib/util/widget_size.dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

// To get a specifc Widget's Size
class WidgetSize extends StatefulWidget {
  final Widget child;
  final Function onChange;

  const WidgetSize({
    Key key,
    @required this.onChange,
    @required this.child,
  }) : super(key: key);

  @override
  _WidgetSizeState createState() => _WidgetSizeState();
}

class _WidgetSizeState extends State<WidgetSize> {
  @override
  Widget build(BuildContext context) {
    SchedulerBinding.instance.addPostFrameCallback(postFrameCallback);
    return Container(
      key: widgetKey,
      child: widget.child,
    );
  }

  var widgetKey = GlobalKey();
  var oldSize;

  // When Widget is built, it is called
  void postFrameCallback(_) {
    var context = widgetKey.currentContext;
    if (context == null) return;

    var newSize = context.size;
    if (oldSize == newSize) return;

    oldSize = newSize;
    widget.onChange(newSize);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:neonpen/neonpen.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.red,
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 0,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.yellow,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.3,
              disableEmphasis: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.orange,
              padding: EdgeInsets.only(
                top: -10,
                left: 5,
                right: 5,
              ),
              opacity: 0.3,
              disableEmphasis: true,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.grey,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.7,
              emphasisWidth: 5,
              emphasisOpacity: 0.5,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.blue,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.3,
              emphasisWidth: 10,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 2,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.purple,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 15),
              opacity: 0.3,
              emphasisWidth: 10,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 2,
              enableLineZiggle: true,
              lineZiggleLevel: 2,
              isDoubleLayer: false,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.green,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 20),
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.lime,
              padding: EdgeInsets.symmetric(vertical: 5, horizontal: 20),
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 2,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.amber,
              padding: EdgeInsets.symmetric(vertical: 10, horizontal: 25),
              opacity: 0.3,
              emphasisWidth: 5,
              emphasisOpacity: 0.3,
              emphasisAngleDegree: 0.5,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
            Neonpen(
              text: Text(
                'Hello World!',
                style: TextStyle(
                  fontSize: 15,
                ),
              ),
              color: Colors.indigo,
              padding: EdgeInsets.symmetric(vertical: 10, horizontal: 25),
              opacity: 0.3,
              emphasisWidth: 8,
              emphasisOpacity: 0.1,
              emphasisAngleDegree: 0.8,
              enableLineZiggle: true,
              lineZiggleLevel: 1,
              isDoubleLayer: true,
            ),
            SizedBox(height: 30),
          ],
        ),
      ),
    );
  }
}
